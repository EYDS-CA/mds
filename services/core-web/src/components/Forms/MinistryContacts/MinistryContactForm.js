import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { compose } from "redux";
import { Field, getFormValues } from "redux-form";
import { Col, Row } from "antd";
import { CoreTooltip } from "@/components/common/CoreTooltip";
import {
  required,
  email,
  phoneNumber,
  maxLength,
  requiredRadioButton,
} from "@mds/common/redux/utils/Validate";
import {
  getMineRegionDropdownOptions,
  getDropdownMinistryContactTypes,
} from "@mds/common/redux/selectors/staticContentSelectors";
import { normalizePhone } from "@common/utils/helpers";
import * as FORM from "@/constants/forms";
import { renderConfig } from "@/components/common/config";
import CustomPropTypes from "@/customPropTypes";
import FormWrapper from "@mds/common/components/forms/FormWrapper";
import RenderCancelButton from "@mds/common/components/forms/RenderCancelButton";
import RenderSubmitButton from "@mds/common/components/forms/RenderSubmitButton";

const propTypes = {
  onSubmit: PropTypes.func.isRequired,
  initialValues: PropTypes.any,
  regionDropdownOptions: PropTypes.arrayOf(CustomPropTypes.dropdownListItem).isRequired,
  MinistryContactTypes: PropTypes.arrayOf(CustomPropTypes.dropdownListItem).isRequired,
  formValues: PropTypes.objectOf(PropTypes.any),
  isEdit: PropTypes.bool.isRequired,
  title: PropTypes.string.isRequired,
  contacts: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.any)).isRequired,
};

const defaultProps = {
  formValues: {},
};

const regionalOfficeCode = "ROE";
const majorMineOfficeCode = "MMO";
const chiefPermittingCode = "CHP";
const chiefInspectorCode = "CHI";
const officeCodes = [regionalOfficeCode, majorMineOfficeCode];
export const MinistryContactForm = (props) => {
  const filteredContactTypes = () => {
    const codes = [];
    const containsAllOffices =
      props.contacts.filter(
        ({ emli_contact_type_code }) => emli_contact_type_code === regionalOfficeCode
      ).length === props.regionDropdownOptions.length;
    const containsMMO = props.contacts.some(
      ({ emli_contact_type_code }) => emli_contact_type_code === majorMineOfficeCode
    );
    const containsCheifPermitting = props.contacts.some(
      ({ emli_contact_type_code }) => emli_contact_type_code === chiefPermittingCode
    );
    const containsCheifInspector = props.contacts.some(
      ({ emli_contact_type_code }) => emli_contact_type_code === chiefInspectorCode
    );
    if (!props.isEdit) {
      if (containsAllOffices) {
        codes.push(regionalOfficeCode);
      }
      if (containsMMO) {
        codes.push(majorMineOfficeCode);
      }
      if (containsCheifPermitting) {
        codes.push(chiefPermittingCode);
      }
      if (containsCheifInspector) {
        codes.push(chiefInspectorCode);
      }
    }
    return props.MinistryContactTypes.filter(({ value }) => !codes.includes(value));
  };

  return (
    <div>
      <FormWrapper onSubmit={props.onSubmit}
        isModal
        initialValues={props.initialValues}
        name={FORM.MINISTRY_CONTACT_FORM}
        reduxFormConfig={{
          touchOnBlur: false,
          enableReinitialize: true,
        }}
      >
        <Row gutter={16}>
          <Col span={12}>
            <Field
              id="is_major_mine"
              name="is_major_mine"
              label="Is this a Major Mine Contact?"
              type="checkbox"
              component={renderConfig.RADIO}
              required
              validate={[requiredRadioButton]}
              disabled={props.isEdit}
            />
          </Col>
          {props.formValues.is_major_mine && (
            <Col span={12}>
              <Field
                id="is_general_contact"
                name="is_general_contact"
                label={
                  <>
                    Is this a general contact?{" "}
                    <CoreTooltip title="General Contacts will be shown on MineSpace in addition to the Regional Contacts." />
                  </>
                }
                type="checkbox"
                required
                validate={[requiredRadioButton]}
                component={renderConfig.RADIO}
              />
            </Col>
          )}
        </Row>

        <Row gutter={16}>
          <Col md={12} xs={24}>
            <Field
              id="mine_region_code"
              name="mine_region_code"
              label={"Mine Region"}
              placeholder="Select a mine Region"
              component={renderConfig.SELECT}
              required={!props.formValues.is_major_mine}
              validate={
                props.formValues.is_major_mine
                  ? []
                  : [required]
              }
              data={props.regionDropdownOptions}
              disabled={props.isEdit}
            />
          </Col>
          <Col md={12} xs={24}>
            <Field
              id="emli_contact_type_code"
              name="emli_contact_type_code"
              label="Contact Type"
              placeholder="Select a contact type"
              component={renderConfig.SELECT}
              required
              validate={[required]}
              data={filteredContactTypes()}
              disabled={props.isEdit}
            />
          </Col>
        </Row>
        {!officeCodes.includes(props.formValues.emli_contact_type_code) && (
          <Row gutter={16}>
            <Col md={12} xs={24}>
              <Field
                id="first_name"
                name="first_name"
                label="First Name"
                component={renderConfig.FIELD}
                required
                validate={[required]}
                disabled={!props.formValues.emli_contact_type_code}
              />
            </Col>
            <Col md={12} xs={24}>
              <Field
                id="last_name"
                name="last_name"
                label="Surname"
                component={renderConfig.FIELD}
                required
                validate={[required]}
                disabled={!props.formValues.emli_contact_type_code}
              />
            </Col>
          </Row>
        )}

        <Row gutter={16}>
          <Col span={24}>
            <Field
              id="email"
              name="email"
              label="Email"
              component={renderConfig.FIELD}
              required
              validate={[email, required]}
              disabled={!props.formValues.emli_contact_type_code}
            />
          </Col>
        </Row>
        <Row gutter={16}>
          <Col span={24}>
            <Field
              id="phone_number"
              name="phone_number"
              label="Phone Number"
              placeholder="e.g. xxx-xxx-xxxx"
              component={renderConfig.FIELD}
              required
              validate={[required, phoneNumber, maxLength(12)]}
              normalize={normalizePhone}
              disabled={!props.formValues.emli_contact_type_code}
            />
          </Col>
        </Row>
        {officeCodes.includes(props.formValues.emli_contact_type_code) && (
          <>
            <Row gutter={16}>
              <Col span={24}>
                <Field
                  id="fax_number"
                  name="fax_number"
                  label="Fax Number"
                  placeholder="e.g. xxx-xxx-xxxx"
                  component={renderConfig.FIELD}
                  validate={[phoneNumber, maxLength(12)]}
                  normalize={normalizePhone}
                  disabled={!props.formValues.emli_contact_type_code}
                />
              </Col>
            </Row>
            <Row gutter={16}>
              <Col span={24}>
                <Field
                  id="mailing_address_line_1"
                  name="mailing_address_line_1"
                  label={"Mailing Address Line 1"}
                  component={renderConfig.AUTO_SIZE_FIELD}
                  required={props.formValues.emli_contact_type_code === regionalOfficeCode}
                  validate={
                    props.formValues.emli_contact_type_code === regionalOfficeCode
                      ? [required]
                      : []
                  }
                  disabled={!props.formValues.emli_contact_type_code}
                />
              </Col>
            </Row>
            <Row gutter={16}>
              <Col span={24}>
                <Field
                  id="mailing_address_line_2"
                  name="mailing_address_line_2"
                  label={"Mailing Address Line 2"}
                  component={renderConfig.AUTO_SIZE_FIELD}
                  required={props.formValues.emli_contact_type_code === regionalOfficeCode}
                  validate={
                    props.formValues.emli_contact_type_code === regionalOfficeCode
                      ? [required]
                      : []
                  }
                  disabled={!props.formValues.emli_contact_type_code}
                />
              </Col>
            </Row>
          </>
        )}

        <div className="right center-mobile">
          <RenderCancelButton />
          <RenderSubmitButton buttonText={props.title} />
        </div>
      </FormWrapper>
    </div>
  );
};

MinistryContactForm.propTypes = propTypes;
MinistryContactForm.defaultProps = defaultProps;

export default compose(
  connect((state) => ({
    formValues: getFormValues(FORM.MINISTRY_CONTACT_FORM)(state) || {},
    regionDropdownOptions: getMineRegionDropdownOptions(state),
    MinistryContactTypes: getDropdownMinistryContactTypes(state),
  }))
)(MinistryContactForm);
