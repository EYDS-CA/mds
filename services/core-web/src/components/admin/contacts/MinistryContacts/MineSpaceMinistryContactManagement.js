import React, { Component } from "react";
import { Divider, Alert } from "antd";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { AuthorizationGuard } from "@/HOC/AuthorizationGuard";
import { openModal, closeModal } from "@mds/common/redux/actions/modalActions";
import {
  getMineRegionHash,
  getMinistryContactTypesHash,
} from "@mds/common/redux/selectors/staticContentSelectors";
import { getMinistryContacts } from "@mds/common/redux/selectors/minespaceSelector";
import {
  fetchMinistryContacts,
  updateMinistryContact,
  deleteMinistryContact,
  createMinistryContact,
} from "@mds/common/redux/actionCreators/minespaceActionCreator";
import { modalConfig } from "@/components/modalContent/config";
import AuthorizationWrapper from "@/components/common/wrappers/AuthorizationWrapper";
import * as Permission from "@/constants/permissions";
import MinistryContactsTable from "@/components/admin/contacts/MinistryContacts/MinistryContactsTable";
import AddButton from "@/components/common/buttons/AddButton";

const propTypes = {
  fetchMinistryContacts: PropTypes.func.isRequired,
  updateMinistryContact: PropTypes.func.isRequired,
  deleteMinistryContact: PropTypes.func.isRequired,
  createMinistryContact: PropTypes.func.isRequired,
  openModal: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  MinistryContacts: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.any)).isRequired,
  mineRegionHash: PropTypes.objectOf(PropTypes.string).isRequired,
  MinistryContactTypesHash: PropTypes.objectOf(PropTypes.string).isRequired,
};

const defaultProps = {};

export class MineSpaceMinistryContactManagement extends Component {
  state = { isLoaded: false };

  componentWillMount() {
    this.handleFetchMinistryContacts();
  }

  handleFetchMinistryContacts = () => {
    this.props.fetchMinistryContacts().then(() => {
      this.setState({ isLoaded: true });
    });
  };

  handleCreateContact = (values) => {
    this.props.createMinistryContact(values).then(() => {
      this.handleFetchMinistryContacts();
      this.props.closeModal();
    });
  };

  handleUpdateContact = (values) => {
    this.props.updateMinistryContact(values.contact_guid, values).then(() => {
      this.handleFetchMinistryContacts();
      this.props.closeModal();
    });
  };

  handleDeleteContact = (guid) => {
    this.setState({ isLoaded: false });
    this.props.deleteMinistryContact(guid).then(() => {
      this.handleFetchMinistryContacts();
    });
  };

  openContactModal = (isEdit, record = null) => {
    return this.props.openModal({
      props: {
        title: isEdit ? "Update MCM Contact" : "Create MCM Contact",
        closeModal: this.props.closeModal,
        initialValues: isEdit ? record : {},
        onSubmit: isEdit ? this.handleUpdateContact : this.handleCreateContact,
        isEdit,
        contacts: this.props.MinistryContacts,
      },
      content: modalConfig.MINISTRY_CONTACT_MODAL,
    });
  };

  render() {
    const officeCodes = ["ROE", "MMO"];
    const offices = this.props.MinistryContacts.filter(({ emli_contact_type_code }) =>
      officeCodes.includes(emli_contact_type_code)
    );
    const contacts = this.props.MinistryContacts.filter(
      ({ emli_contact_type_code }) => !officeCodes.includes(emli_contact_type_code)
    );
    return (
      <div>
        <div className="landing-page__header">
          <div className="inline-flex between">
            <h1>MineSpace MCM Contact Management</h1>
            <AuthorizationWrapper permission={Permission.ADMIN}>
              <AddButton onClick={() => this.openContactModal(false)}>
                Create MCM Contact
              </AddButton>
            </AuthorizationWrapper>
          </div>
        </div>
        <div className="tab__content">
          <Alert
            message="MCM contacts and offices are displayed in multiple places within Core and MineSpace."
            closable
            description={
              <>
                The Contacts are displayed in MineSpace to proponents.
                <br />
                The Offices are displayed in MineSpace, shown on the footer on permits, and NoW
                documents, and are the default email address for variances, code required reports,
                and mine information notifications.
              </>
            }
            type="info"
            showIcon
          />
          <h2>Offices</h2>
          <Divider />
          <MinistryContactsTable
            isLoaded={this.state.isLoaded}
            contacts={offices}
            isOffice
            mineRegionHash={this.props.mineRegionHash}
            openEditModal={this.openContactModal}
            handleDeleteContact={this.handleDeleteContact}
            MinistryContactTypesHash={this.props.MinistryContactTypesHash}
          />
          <br />
          <h2>Contacts</h2>
          <Divider />
          <MinistryContactsTable
            isLoaded={this.state.isLoaded}
            contacts={contacts}
            mineRegionHash={this.props.mineRegionHash}
            openEditModal={this.openContactModal}
            handleDeleteContact={this.handleDeleteContact}
            MinistryContactTypesHash={this.props.MinistryContactTypesHash}
          />
        </div>
      </div>
    );
  }
}

MineSpaceMinistryContactManagement.propTypes = propTypes;
MineSpaceMinistryContactManagement.defaultProps = defaultProps;

const mapStateToProps = (state) => ({
  MinistryContacts: getMinistryContacts(state),
  mineRegionHash: getMineRegionHash(state),
  MinistryContactTypesHash: getMinistryContactTypesHash(state),
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      fetchMinistryContacts,
      updateMinistryContact,
      deleteMinistryContact,
      createMinistryContact,
      openModal,
      closeModal,
    },
    dispatch
  );

export default AuthorizationGuard(Permission.EDIT_MINISTRY_CONTACTS)(
  connect(mapStateToProps, mapDispatchToProps)(MineSpaceMinistryContactManagement)
);
