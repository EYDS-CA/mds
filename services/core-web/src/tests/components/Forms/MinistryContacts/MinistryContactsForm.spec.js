import React from "react";
import { shallow } from "enzyme";
import { MinistryContactForm } from "@/components/Forms/MinistryContacts/MinistryContactForm";

const props = {};

const setupProps = () => {
  props.submitting = false;
  props.title = "Update MCM Contact";
  props.handleSubmit = jest.fn();
  props.closeModal = jest.fn();
  props.regionDropdownOptions = [];
  props.MinistryContactTypes = [];
  props.isEdit = true;
  props.initialValues = {};
  props.formValues = {};
  props.contacts = [];
};

beforeEach(() => {
  setupProps();
});

describe("MinistryContactForm", () => {
  it("renders properly", () => {
    const component = shallow(<MinistryContactForm {...props} />);
    expect(component).toMatchSnapshot();
  });
});
