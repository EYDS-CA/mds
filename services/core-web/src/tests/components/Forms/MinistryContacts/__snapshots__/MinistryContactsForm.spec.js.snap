// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MinistryContactForm renders properly 1`] = `
<div>
  <Connect(ReduxForm)
    initialValues={Object {}}
    isModal={true}
    reduxFormConfig={
      Object {
        "enableReinitialize": true,
        "touchOnBlur": false,
      }
    }
  >
    <Row
      gutter={16}
    >
      <Col
        span={12}
      >
        <Field
          component={[Function]}
          disabled={true}
          id="is_major_mine"
          label="Is this a Major Mine Contact?"
          name="is_major_mine"
          required={true}
          type="checkbox"
          validate={
            Array [
              [Function],
            ]
          }
        />
      </Col>
    </Row>
    <Row
      gutter={16}
    >
      <Col
        md={12}
        xs={24}
      >
        <Field
          component={[Function]}
          data={Array []}
          disabled={true}
          id="mine_region_code"
          label="Mine Region"
          name="mine_region_code"
          placeholder="Select a mine Region"
          required={true}
          validate={
            Array [
              [Function],
            ]
          }
        />
      </Col>
      <Col
        md={12}
        xs={24}
      >
        <Field
          component={[Function]}
          data={Array []}
          disabled={true}
          id="emli_contact_type_code"
          label="Contact Type"
          name="emli_contact_type_code"
          placeholder="Select a contact type"
          required={true}
          validate={
            Array [
              [Function],
            ]
          }
        />
      </Col>
    </Row>
    <Row
      gutter={16}
    >
      <Col
        md={12}
        xs={24}
      >
        <Field
          component={[Function]}
          disabled={true}
          id="first_name"
          label="First Name"
          name="first_name"
          required={true}
          validate={
            Array [
              [Function],
            ]
          }
        />
      </Col>
      <Col
        md={12}
        xs={24}
      >
        <Field
          component={[Function]}
          disabled={true}
          id="last_name"
          label="Surname"
          name="last_name"
          required={true}
          validate={
            Array [
              [Function],
            ]
          }
        />
      </Col>
    </Row>
    <Row
      gutter={16}
    >
      <Col
        span={24}
      >
        <Field
          component={[Function]}
          disabled={true}
          id="email"
          label="Email"
          name="email"
          required={true}
          validate={
            Array [
              [Function],
              [Function],
            ]
          }
        />
      </Col>
    </Row>
    <Row
      gutter={16}
    >
      <Col
        span={24}
      >
        <Field
          component={[Function]}
          disabled={true}
          id="phone_number"
          label="Phone Number"
          name="phone_number"
          normalize={[Function]}
          placeholder="e.g. xxx-xxx-xxxx"
          required={true}
          validate={
            Array [
              [Function],
              [Function],
              [Function],
            ]
          }
        />
      </Col>
    </Row>
    <div
      className="right center-mobile"
    >
      <RenderCancelButton />
      <RenderSubmitButton
        buttonText="Update MCM Contact"
      />
    </div>
  </Connect(ReduxForm)>
</div>
`;
